extensions/notebook/src/book/bookTocManager.ts

```diff
@@ -0,0 +1,181 @@
+	createBook(bookContentPath: string, contentFolder: string): Promise<void>
+}
+const allowedFileExtensions: string[] = ['.md', '.ipynb'];
+const initMarkdown: string[] = ['index.md', 'introduction.md', 'intro.md', 'readme.md'];
+
+export function hasSections(node: JupyterBookSection): boolean {
+	return node.sections !== undefined && node.sections.length > 0;
+}
+
+export class BookTocManager implements IBookTocManager {
+	public tableofContents: IJupyterBookSectionV2[];
+	public newSection: JupyterBookSection = {};
+
+	constructor() {
+	}
+
+	async getAllFiles(toc: IJupyterBookSectionV2[], directory: string, filesInDir: string[], rootDirectory: string): Promise<IJupyterBookSectionV2[]> {
+		await Promise.all(filesInDir.map(async file => {
+			let isDirectory = (await fs.promises.stat(path.join(directory, file))).isDirectory();
+			if (isDirectory) {
+				let files = await fs.promises.readdir(path.join(directory, file));
+				let initFile: string = '';
+				//Add files named as readme or index whithin the directory as the first file of the section.
+				files.some((f, index) => {
+					if (initMarkdown.includes(f)) {
+						initFile = path.parse(f).name;
+						files.splice(index, 1);
+					}
+				});
+				let jupyterSection: IJupyterBookSectionV2 = {
+					title: file,
+					file: path.join(file, initFile),
+					expand_sections: true,
+					numbered: false,
+					sections: []
+				};
+				toc.push(jupyterSection);
+				await this.getAllFiles(toc, path.join(directory, file), files, rootDirectory);
+			} else if (allowedFileExtensions.includes(path.extname(file))) {
+				// if the file is in the book root we don't include the directory.
+				const filePath = directory === rootDirectory ? path.parse(file).name : path.join(path.basename(directory), path.parse(file).name);
+				const addFile: IJupyterBookSectionV2 = {
+					title: path.parse(file).name,
+					file: filePath
+				};
+				//find if the directory (section) of the file exists else just add the file at the end of the table of contents
+				let indexToc = toc.findIndex(parent => parent.title === path.basename(directory));
+				//if there is not init markdown file then add the first notebook or markdown file that is found
+				if (indexToc !== -1) {
+					if (toc[indexToc].file === '') {
+						toc[indexToc].file = addFile.file;
+					} else {
+						toc[indexToc].sections.push(addFile);
+					}
+				} else {
+					toc.push(addFile);
+				}
+			}
+		}));
+		return toc;
+	}
+
+	updateToc(tableOfContents: JupyterBookSection[], findSection: BookTreeItem, addSection: JupyterBookSection): JupyterBookSection[] {
+		for (const section of tableOfContents) {
+			if ((section as IJupyterBookSectionV1).url && path.dirname(section.url) === path.join(path.sep, path.dirname(findSection.uri)) || (section as IJupyterBookSectionV2).file && path.dirname((section as IJupyterBookSectionV2).file) === path.join(path.sep, path.dirname(findSection.uri))) {
```

Comment:
same question as below

Reply:
Yeah, some xml docs on this or comment would help me better understand.

Reply:
The tableOfContents contains an array of all the sections and I recursively look for the section of the book that's being modified. In the table of contents, we get the information in one of these two formats:

title: Troubleshooters
url: /troubleshooters/readme
-OR-

title: Troubleshooters
file: /troubleshooters/readme
findSection is a bookTreeItem and contains the path to the first url of the section in the uri property, but it's stored as troubleshooters/readme

That's why when comparing I get the directory name for both and append a path.sep to the findSection uri